### PR TITLE
feat: add v5 for Internal segment

### PR DESCRIPTION
**Pull request by JackTan25**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12348)

### CODE REVIEW

File: src/query/storages/common/table-meta/src/meta/v5/segment.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+            + segments_compress.len()
+            + summary_compress.len().to_le_bytes().len()
+            + summary_compress.len();
+        let mut buf = Vec::with_capacity(data_size);
+
+        buf.extend_from_slice(&self.format_version.to_le_bytes());
+        buf.push(encoding as u8);
+        buf.push(compression as u8);
+        buf.extend_from_slice(&segments_compress.len().to_le_bytes());
+        buf.extend_from_slice(&segments_compress.len().to_le_bytes());
```
Comment (dantengsky) [inline src/query/storages/common/table-meta/src/meta/v5/segment.rs:91]: typo?

File: src/query/storages/common/table-meta/src/meta/v5/segment.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            segment = Self::new_leaf(blocks, summary);
+        } else if version == Self::VERSION {
+            let son_segments: Vec<Arc<Location>> =
+                read_and_deserialize(&mut cursor, blocks_size, &encoding, &compression)?;
+            let summary: Statistics =
+                read_and_deserialize(&mut cursor, summary_size, &encoding, &compression)?;
+
+            segment = Self::new_internal(son_segments, summary);
+        } else {
+            panic!("read bad version for v5")
```
Comment (dantengsky) [inline src/query/storages/common/table-meta/src/meta/v5/segment.rs:200]: I might understand incorrectly, but how about making the `pub enum SegmentInfo` `Serialize` and `Deserialize`

File: src/query/storages/common/table-meta/src/meta/v5/segment.rs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    ///
+    ///   e.g. if the current version of SegmentInfo is v3::SegmentInfo, then the format_version
+    ///   that will be written down to object storage as part of SegmentInfo table meta data,
+    ///   should always be v3::SegmentInfo::VERSION (which is 3)
+    pub format_version: FormatVersion,
+    /// segments belong to this internal_segment
+    /// there will be normal segments and internal segment
+    /// the suffix name of different segments is different,
+    /// so we can use this tag to distinct them.
+    pub son_segments: Vec<Arc<Location>>,
```
[the remark this answers is not in the document]
Comment (ariesdevil) [reply src/query/storages/common/table-meta/src/meta/v5/segment.rs:60]: The prefix `son` is rarely used in variable naming, consider using child instead.

[the remark this answers is not in the document]
Comment (JackTan25) [reply src/query/storages/common/table-meta/src/meta/v5/segment.rs:60]: > why boxed with Arc? any concerns?

no much concerns, I reference to https://github.com/datafuselabs/databend/blob/main/src/query/storages/common/table-meta/src/meta/v4/segment.rs#L58 here, the blocks in v4:SegmentInfo is Arc?

File: src/query/storages/common/table-meta/src/meta/v5/segment.rs (hunk 2 above)
Comment (JackTan25) [reply src/query/storages/common/table-meta/src/meta/v5/segment.rs:200]: Just use to_bytes and from_slice, is that incorrect?

File: src/query/storages/common/table-meta/src/meta/v5/segment.rs (hunk 1 above)
Comment (JackTan25) [reply src/query/storages/common/table-meta/src/meta/v5/segment.rs]: > typo?

fixed

File: src/query/storages/common/table-meta/src/meta/v5/segment.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    InternalSegment(InternalSegmentInfo),
+    LeafSegment(LeafSegmentInfo),
+}
+
+impl From<v3::SegmentInfo> for SegmentInfo {
+    fn from(value: v3::SegmentInfo) -> Self {
+        SegmentInfo::LeafSegment(LeafSegmentInfo::from_v3(value))
+    }
+}
+
```
Comment (zhyass) [inline src/query/storages/common/table-meta/src/meta/v5/segment.rs:115]: Need add 
```
impl From<v4::SegmentInfo> for SegmentInfo {
    fn from(value: v4::SegmentInfo) -> Self {
        SegmentInfo::LeafSegment(value)
    }
}
```

File: src/query/storages/common/table-meta/src/meta/v5/segment.rs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        let SegmentHeader {
+            version,
+            encoding,
+            compression,
+            // maybe we need to modify as blocks_or_segments_size
+            blocks_size,
+            summary_size,
+        } = decode_segment_header(&mut cursor)?;
+
+        if version == LeafSegmentInfo::VERSION {
```
Comment (zhyass) [inline src/query/storages/common/table-meta/src/meta/v5/segment.rs:186]: Is it reasonable to use VERSION to determine whether it is LeafSegmentInfo or InternalSegment?
Maybe we need add a flag in SegmentHeader to determine when serialize and  deserialize.

Comment (JackTan25) [reply src/query/storages/common/table-meta/src/meta/v5/segment.rs]: fixed, no need to change SegmentHeader struct, we can get it outside

File: src/query/storages/common/table-meta/src/meta/v5/segment.rs (hunk 2 above)
Comment (JackTan25) [reply src/query/storages/common/table-meta/src/meta/v5/segment.rs]: Maybe I misunderstood your meaning, I add *#[derive(Serialize, Deserialize)]* for *pub enum SegmentInfo*, and we will use from_slice and to_bytes.

**Review by zhyass — APPROVED** [review]

**Review by dantengsky — APPROVED** [review]
Let merge, and refine step by step

**Review by lichuang — APPROVED** [review]

**Pull request merged**